### PR TITLE
Avoid memory copy for non-redundant blocks scrub.

### PR DESCRIPTION
**Pull request by amotin**
When scrubbed block has several copies (ditto or mirror), we allocate
separate buffer for each one and then copy successfully read data to
the parent ZIO buffer.  But if there is only one copy, then it makes
no sense to do the alloc/../copy/free and we can just use the parent
ZIO buffer directly.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Review by behlendorf — APPROVED** [review]
If I had known you were also looking at this area of the code I would have tagged you for a review for ad8b9f940c1e39a38af61934737c1e4cf8ab5c08!  This looks good to me, although the logic of where those new ABDs get freed (while correct) is a bit hard to follow!

**Comment by amotin** [comment]
I wasn't in this area (vdev_mirror) in particular, just spotted strange memcpy on a scrub CPU profiles I was cleaning for the past couple weeks.  I detected the collision and went to look only when you've merged your commit.  A complete coincidence, but a quite precise one indeed. :)

**Review by mmaybee — COMMENTED** [review]
This seems like a lot of complexity for the odd case of a single child in a mirror IO... but couldn't we generalize this concept for multiple children? Take the `zio_abd` and assign it to the first available child and create new abd structs for the remaining children. Then, in the done func, assign the "first" child's abd struct back to the zio and clean up the remaining and structs.

**Comment by amotin** [comment]
@mmaybee It is not an odd case.  Root vdev is actually a mirror inside, handling ditto blocks with a mirror logic.  And most of blocks have only one copy, that is why it is important.

What's about proposition of using parent ZIO buffer for the first child ZIO of actual mirrors, in that case vdev_mirror_scrub_done() for later ZIOs could not handle the data copy if the first ZIO is still in progress. The data copy would have to be moved into vdev_mirror_io_done(), which up to the recent @behlendorf commit had no idea about the children ZIO buffers.  If somebody wish to refactor it that way now when it has, it could have sense.

**Comment by amotin** [comment]
> If somebody wish to refactor it

Nevermind, I'll refactor it together with https://github.com/openzfs/zfs/pull/13593. I too much like idea of not having any memory copies, only check-summing. :)
